geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HKeysDUnitTest.java

```diff
@@ -0,0 +1,165 @@
+  private static Jedis jedis2;
+
+  private static Properties locatorProperties;
+
+  private static MemberVM locator;
+  private static MemberVM server1;
+  private static MemberVM server2;
+
+  private static int redisServerPort1;
+  private static int redisServerPort2;
+
+  @BeforeClass
+  public static void classSetup() {
+    locatorProperties = new Properties();
+
+    locator = clusterStartUp.startLocatorVM(0, locatorProperties);
+    server1 = clusterStartUp.startRedisVM(1, locator.getPort());
+    server2 = clusterStartUp.startRedisVM(2, locator.getPort());
+
+    redisServerPort1 = clusterStartUp.getRedisPort(1);
+    redisServerPort2 = clusterStartUp.getRedisPort(2);
+
+    jedis1 = new Jedis(LOCAL_HOST, redisServerPort1, JEDIS_TIMEOUT);
+    jedis2 = new Jedis(LOCAL_HOST, redisServerPort2, JEDIS_TIMEOUT);
+  }
+
+  @Before
+  public void testSetup() {
+    jedis1.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis1.disconnect();
+    jedis2.disconnect();
+
+    server1.stop();
+    server2.stop();
+    locator.stop();
+  }
+
+  @Test
+  public void testConcurrentHKeys_whileAddingValues() {
+    String key = "key";
+
+    Map<String, String> testMap = makeHashMap(HASH_SIZE, "field-", "value-");
+    Set<String> expectedFields = makeSet(HASH_SIZE, "field-");
+    BlockingQueue<String> queue = new LinkedBlockingQueue<>();
+
+    jedis1.hset(key, testMap);
+
+    new ConcurrentLoopingThreads(NUM_ITERATIONS,
+        (i) -> {
+          int newIndex = HASH_SIZE + i + 1;
+          jedis1.hset(key, "field-" + newIndex, "value-" + newIndex);
+          queue.add("field-" + newIndex);
+        },
+        (i) -> {
+          try {
+            expectedFields.add(queue.take());
+          } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+          }
+          assertThat(jedis2.hkeys(key)).containsAll(expectedFields);
+        }).run();
+
+    assertThat(jedis1.hkeys(key)).containsExactlyInAnyOrderElementsOf(expectedFields);
+  }
+
+  @Test
+  public void testConcurrentHKeys_whileDeletingValues() {
+    String key = "key";
+    Map<String, String> testMap = makeHashMap(NUM_ITERATIONS, "field-", "value-");
+
+    jedis1.hset(key, testMap);
+
+    new ConcurrentLoopingThreads(NUM_ITERATIONS,
+        (i) -> jedis1.hdel(key, "field-" + i),
+        (i) -> jedis2.hkeys(key)).run();
+
+    assertThat(jedis1.hkeys(key).size()).isEqualTo(0);
+  }
+
+  @Test
+  public void testConcurrentHKeys_whileUpdatingValues() {
+    String key = "key";
+    Map<String, String> testMap = makeHashMap(NUM_ITERATIONS, "field-", "value-");
+
+    jedis1.hset(key, testMap);
+
+    new ConcurrentLoopingThreads(NUM_ITERATIONS,
+        (i) -> jedis1.hset(key, "field-" + i, "changedValue-" + i),
+        (i) -> assertThat(jedis2.hkeys(key)).containsExactlyInAnyOrderElementsOf(testMap.keySet()))
+            .run();
+
+    assertThat(jedis1.hkeys(key)).containsExactlyInAnyOrderElementsOf(testMap.keySet());
```

Comment:
Technically, this test doesn't confirm that the values were actually changed by the thread on line 142. It just confirms the keys are all still there. A loop or assert checking that the new values are all "changedValue-[number]" might be useful.

Reply:
👍